### ISSUE TITLE
Is it possible, Could we develop a mass matrix at each time step in the case  {(ε + 3 chi | E |^² I  + 2 Chi ^3 A)} dE/dt?

### ISSUE THREAD

**Issue by Asadanees** [mfem/mfem#200]
Dear @Dobrev and @Tzanio Kolev
I have developed MFEM codes and error estimates analytically  for Linear Maxwell’s equation. 
Also, I have developed analytically error estimates for Nonlinear Electromagnetic and Optics. 
Now I am extending this code for Nonlinear Optics and Electromagnetic.
I studied  ex10 and ex16, where u is s scalar, in ex16 matrix K is recreated on the next Implicit solve.
However, in our case E (electric field) and H (Magnetic Field) are Vectors unknown.
How we can construct a Mass Matrix ?

                                         ( ε + 3 chi | E|² ) dE/dt  =   Curl x H,   where ε  is a constant and chi > 0
					          dH/dt = - Curl x E
In my opinion, first I would have to create a VectorGridfunction and second Mass Matrix would need to be  created at each time step. Third, then at each time step invert the Mass Matrix using the Krylov solver.
I need some help to create Mass Matrices and Vectorgridfuction.
Further, I would have to extend this code to general nonlinear Electromagnetics and Optics with Complex geometries. And Finally  parallel code would also be developed for  this non-linear model to improve efficiency specially solver (MPI and OpenMP )
Scientific Visualization is also a part of this project.
I would like to share all my project work with MFEM (LLNL) after publications. Feel free to contact if you have any question. 
Thanks and best wishes for providing the awesome library. @ @ @

**Comment by Asadanees** [comment]
Dear @Dobrev and @Tzanio Kolev
sorry for inconvenience. I wrote Mass Matrix instead of Non-linear Matrix. 
Now I am extending this code for Nonlinear Optics and Electromagnetic.
I studied ex10 and ex16, where u is s scalar, in ex16 matrix K is recreated on the next Implicit solve.
However, in our case E (electric field) and H (Magnetic Field) are Vectors unknown.
How we can construct a Nonlinear Matrix ?

       ( ε + 3 chi | E|²  )  dE/dt  =   Curl x H,   where ε  is a constant and chi > 0
				    dH/dt = - Curl x E
we can also write the first equation
     dE/dt  =  ( ε + 3 chi | E|²  )^{-1}(Curl x H),   where ε  is a constant and chi > 0

In my opinion, first I would have to create a VectorGridfunction and second non-linear Matrix would need to be created at each time step. 
I need some help to create Nonlinear Matrices .
Further, I would have to extend this code to general nonlinear Electromagnetics and Optics with Complex geometries. And Finally parallel code would also be developed for this non-linear model to improve efficiency specially solver (MPI and OpenMP )
Scientific Visualization is also a part of this project.
I would like to solve in both way.
Thanks and best wishes for providing the awesome library. @ @ @

**Comment by mlstowell** [comment]
Hello, Asad,

One of our team members, Aaron Fisher, did a lot of work with this type of non-linearity some time ago.  The scheme he used involved recomputing the primary mass matrix at each time step.  This has the effect of linearizing the non-linear operator.  This is computationally expensive but unavoidable.  Fortunately the sparsity pattern of this operator should not change during time evolution which means that we can allocate the matrix once and simply update its values.  This should improve efficiency somewhat, although computing the matrix entries themselves will probably continue to dominate. 

Aaron's paper on this should be available at this link https://digital.library.unt.edu/ark:/67531/metadc897483/m2/1/high_res_d/923114.pdf.
The title is "An Efficient Vector Finite Element Method for Nonlinear Electromagnetic Modeling".  He also suggests using a symplectic integration technique with a carefully chosen time step and in my experience this is very important.

Please have a look at this paper, particularly sections 2 and 3.3.  This should help you decide exactly which operators will need to be rebuilt at each time step.  Once we have the algorithm defined I'll be happy to help you design some MFEM specific code to implement the matrix update.

Best wishes,
Mark

**Comment by acfisher** [comment]
It has certainly been a while since I have looked at that work.  I'm happy to chime in with any of the tricks I pulled to get some of the big photonic crystal fiber simulations to run on the cluster I had access to at the time.  One caveat that I didn't see mentioned above was that I used a finite response time model for chi which handles the non-linearity in a more time step independent way.  A simpler approach is to just use the last time step value for E in the chi calculation, but it comes with the downside that your response time is now tied to the time step.

**Comment by Asadanees** [comment]
Dear @mlstowell  @acfisher and @tzanio .
Tanks for your cooperation for providing help to design MFEM specific cods for nonlinear Electromagnetics. LLNL scientists already (White, Fisher and Stowell.) done  a awesome work in electromagnetics. I am really sorry for delay. Last month I was at my business tours and also did not clear  one term in he algorithm to find r_{out}(Fisher work 2007).
I tried my best to understand this term from the lecture of White, Rieben, Castillo, Fisher and Stowell. I am really sorry that I could not fix it absolutely. Could you please explain the term in algorithm to compute r_{out}( e_{in}^(12)*M2^(-1)H12)?.   Still, the e_{in}^12 notation is not clear to me.  
Below is the detail of Algorithms for nonlinear Electromagnetics. 

   d/dt (D)=    curl H 
     d/dt (B) = -  curl E                            system (1)
D = eps_{0} (eps(r) + chi^{3} E.E ) 
B = mu_{0} H
the above mathematically model is same as Fisher 2007
nabla x H = curl H
nabla xE =curl E  

eff=eps_{0}(chi^{L} + chi^{NL})
we can write d\dt (D)
d/dt (D)= (eps_{NL} * I_{3x3}+ 2 Chi ^{3} A) d/dt (E)

I_{3x3} is 3 x 3 unit Matrix (Identity unit 3x3 matrix)

 A is  3x3 matrix , the entries of matrix are given below 
         E1.E1 E1.E2    E1.E3    ... first row of A
 A  =  E1.E2   E2.E2  E2.E3    ...second row  of A
         E1.E3   E3.E2   E3.E3    ...third row of A

then system (1) implies that 

(eps_{NL} * I_{3x3}+ 2 Chi ^{3} A) dE/dt=nabla x H
                                                          dH/dt=-nabla x E
 In the article of fisher 2007 , he adopted below approach page 1333 equation (5)

eps_{eff} d\dt E + d/dt( eps_{eff}) E = curl B ….. (5)

in  our techniques we  have no d/dt( eps_{eff})E term , so that in each time step we have no term  M1_{d_eff} e_{in} for e_{out}. 



Debye time relaxation for nonlinear part ( same as fisher work)

 tau d\dt( chi^{NL} ) + chi^{NL} = chi^(3) E.E  

e^{tau *t} chi ^{NL} = \int_\Omega e^{ tau *s} chi ^(3) E.E ds 




M1 is a first form Mass matrix
M1 = \int_\Omega phi_{i}_{1}  phi_{j}_{1} d\Omega
M1^(-1) is the inverse of mass Matrix M1

M2 is s second form Mass matrix
M2=  \int_\Omega phi_{i}_{2} phi_{j}_{2} d\Omega
M2^(-1) is the inverse of mass Matrix M2

D12 is discrete representation of curl
D12=  \int_\Omega (nabla x phi_{i}_{1}) phi_{j}_{2} d\Omega

D12^(T) is a transpose of D12

Hodge Matrix

H12 \int_\Omega  phi_{i}_{1} phi_{j}_{2} d\Omega



 Kth order symplectic time Integration method

nstep= (T_{final} – t_{0}) / (time_step)
 Initial condition

r_{1}  < ----- r_{init}
e_{1} < ------  e_{init}
h_{1} < ------h_{init}

begin loop over the time step 

for i=1    to nstep do

r_{in} < -----  r_{i}

e_{in} < ----  e_{i}

h_{in} < ----- h_{i}


alpha and Beta are constant 
Order 1 
alpha_{1}= 1 and Beta_{1}=1
Order 2
alpha_{1}= 1/2 and 		Beta_{1}=0
alpha_{2}= 1 /2 and 		Beta_{2}=1
for j=1 to order do
r_{out}< ---- r_{in} + Beta_{j}*time_step (r_{in}/tau + chi^(3)\tau (e_{in}^(12)*M2^(-1)H12 e_{in} )
eff < ---- eps +eps_{0} r_{out}
the step need for fisher method only 
d_eff < ---- (r_{out} -r_{in})/ Beta_{j}*time_step   ……...(A)
e_{out} < ---- e_{in}+Beta_{j}*time_step (M1_{eff}^{-1} * D12 ^(T) h_{in})
if we follow the technique of Fisher than we have one more term M1_{d_eff}e_{in} in e_{out}
e_{out} < -- e_{in}+Beta_{j}*time_step*M1_{eff}^{-1} (D12 ^(T) h_{in}-M1_{d_eff}e_{in})….(B)

The steps (A) and (B) evaluate if we adopt the technique of Fisher 2007

h_{out} < ---- h_{in} + alpha_{j} *time_step(-  M2^(-1) *D12 e_{out})
 

r_{in} < ---- r_{out}

e_{in} < --- e_{out}

h_{in} < --- h_{out}

end for j

r_{i+1} < ----- r_{out}

e_{i+1} < ---- e_{out}

h_{i+1} < ---- h_{out}

end for 

e_{final} < ---- e_{nstep+1}
h_{final} < ---- h_{nstep +1}

We will  write the MFEM special code for both techniques (Fisher and us). These MFEM code will be extended in MpI and OpenMp with complex geometries.  Could you please inform me?   could  I email you a PDF file to  our Mathematical result for non-linear electromagnetic ( error estimate)?  If I would like that I could publish my research work with LLNL (FISHER and Stowell) jointly. Feel free to contact if you have any question. 
Thanks and best wishes tzanio @acfisher  and @mlstowell .

**Comment by acfisher** [comment]
Hello Asad,

The ^ in this case is a wedge product (not to be confused with a superscript/exponentiation operation), and the (12) indicates that it is operating on a 1-for and a 2-form.  However in equation 16 I mention that this can be simplified and an extra matrix solve can be avoided by simply projecting the E dot E results onto a 3-form and using that in the calculation instead.  This is what I would recommend.

**Comment by mlstowell** [comment]
Hello, Asad,

The notation certainly does require some time to absorb.  Aaron Fisher and I were trained to think of electromagnetics in the language of discrete differential forms.  MFEM does not use typically use this language but it has all of the same basis functions.

In Aaron's paper E and H would be 1-forms, B and D would be 2-forms, and r is a 3-form.  In MFEM we would represent 1-forms using Nedelec basis functions (indicated by "ND" in the class names), 2-forms using Raviart-Thomas basis functions (indicated by "RT" in the class names), and 3-forms using L2 basis functions.

The symbol ^(12) is meant to represent a "wedge" product (also called an exterior product) between a 1-form and a 2-form.  Such wedge products produce a differential form of a higher type, in this case a 3-form.  Another way to look at this is that a 1-form is a regular vector field, a 2-form is a pseudo-vector field, and their wedge product is just the usual dot product which produces a pseudo-scalar field.

I'm not sure how clear this is but the important thing to understand is that the following discrete formula is just a fancy way to compute dot product of the electric field with itself:
e_{in}^(12)*M2^(-1)H12 e_{in}
Let me try to decode this one piece at a time.
1) e_{in} is a vector of 1-form (Nedelec) degrees of freedom describing the electric field.
2) H12 e_{in} is a vector of integrals of the Raviart-Thomas basis functions multiplied by a discrete approximation of the electric field.
3) M2^(-1)H12 e_{in} is the result of solving the Raviart-Thomas mass matrix using the previous vector as the right hand side.  Consequently this produces a vector of Raviart-Thomas degrees of freedom which describe the same electric field as e_{in} describes but using different basis functions.  This combination of operators is intended to produce a Raviart-Thomas approximation which is the closest fit (in the least squares sense) to the Nedelec approximation.
4) Finally, e_{in}^(12)*(the previous vector) computes the L2 degrees of freedom for the scalar field which approximates the dot product E.E.

You may well ask why we don't simply interpolate the electric field vectors and compute the dot products in the usual manner and project these results onto the L2 basis functions.  This is a reasonable question and I believe many codes use this technique in similar contexts.  I like to think that our method is more accurate but I'm not sure I could prove that, perhaps Aaron has worked out those details.  In any case, the interpolation technique would require evaluating and/or storing Jacobians at every interpolation point throughout the mesh and this has a non-trivial cost.  In our method we can precompute all of the relevant Jacobians and incorporate them into the linear operators which are simple and reasonably inexpensive to apply.  I honestly can't say which technique would be more efficient but I believe ours was more convenient at the time.

(I just noticed Aaron's reply but I hope this will also be useful.)

Best wishes,
Mark

**Comment by mlstowell** [comment]
Hello, Asad,

One other issue I forgot to address.  Your offer to publish jointly is very generous.  However, I don't expect I would be able to make any substantive contribution beyond offering advice in using MFEM for such problems.  I am always happy to answer any MFEM related questions you might have.  If you could reference "http://mfem.org" in any relevant publications I'm sure the team would appreciate that.

Best wishes,
Mark

**Comment by acfisher** [comment]
Thanks for chiming in as well Mark.  I was in a hurry and put out my quick explanation, but you took the time a put some real depth in there.  I'm sure it will be useful.

Cheers,
Aaron

**Comment by tzanio** [comment]
Hi @Asadanees, I hope the @mlstowell's and @acfisher's answers have been useful. I'm closing this issue now. Feel free to reopen if you have any further questions.

**Comment by Asadanees** [comment]
@mlstowell  and @acfisher 
thanks for your cooperations. I want to develop below mass matrix at each time step
( ε + 3 chi | E|² ) dE/dt
This simple approach to develop the mass matrix at each time . 
could you please tell me? is it right approch?

void Maxwell::SetParameters(double _dt, const Vector *_E, const Vector *_H)
{

	
	current_dt=_dt;
	
	H=_H;
	
	E=_E;

        delete N;
	N= new BilinearForm(&Nedelec);
	
	GridFunction E_permittivity_gf(&Nedelec);
	
	
	
	E_permittivity_gf.SetFromTrueDofs(*E);
	
	for(int i=0; i<E_permittivity_gf.Size(); i++)
	{
		
		E_permittivity_gf(i)= rmittivity+ N->InnerProduct(*E,*E)*(E_permittivity_gf(i));
		
	}
	
	
	
	
	
	GridFunctionCoefficient E_coeff(&E_permittivity_gf);
	
	N->AddDomainIntegrator(new VectorFEMassIntegrator(E_coeff));
	
	N->Assemble();
	
}

**Comment by Asadanees** [comment]
dear @tzanio 
please open this issue.
I did not know . How can I open this issue.
Thanks and best wishes.

**Comment by mlstowell** [comment]
Hello, Asad,

There are many possible ways to implement this and I imagine they each have pros and cons in terms of efficiency and ability to work well in parallel, etc..

That being said, the pseudo code you provide looks like a reasonable start but it does have three obvious issues.  First, the BilinearForm N cannot be used to compute the inner product as written.  Second, the inner product should be computed before the loop because its value should not change. Lastly, the GridFunction and GridFucntionCoefficient will be deleted when they go out of scope at the end of this method which might be dangerous.

Issue 1: The BilinearForm used to compute the InnerProduct needs to be assembled before this method can be called.  I imagine it could be reused each time this method is called so perhaps you could declare a second BilinearForm named N0, for example.

// Somewhere outside the SetParameters method
N0 = new BilinearForm(&Nedelec);
N0->AddDomainIntegrator(new VectorFEMassIntegrator());
N0->Assemble();

Issue 2:
// Then in SetParameters N0 is available to compute inner products
double E2 = N0->InnerProduct(*E,*E);

GridFunction * E_permittivity_gf = new GridFunction(&Nedelec);
*E_permittivity_gf = rmittivity;
E_permittivity_gf->Add(E2, *E);

Issue 3:
The GridFunction and GridFunctionCoefficient could be data members of the Maxwell class so that they persist outside this method.

**Comment by Asadanees** [comment]
Dear @mlstowell,
Thank for your correction and cooperation. it is nice. 
Now I want to extend this code general type of non-linearity 
I did not know how to contract the mass matrix at each time step.  

{(ε + 3 chi | E|² * I_{3x3}+ 2 *Chi ^{3}* A)} dE/dt

I_{3x3} is 3 x 3 unit Matrix (Identity unit 3x3 matrix)

A is 3x3 matrix , the entries of matrix are given below

            E1.E1     E1.E2     E1.E3         ... first row of A
     A =  E1.E2     E2.E2     E2.E3         ...second row of A
            E1.E3      E3.E2     E3.E3        ...third row of A

after this development  I will use the approach @acfisher (an efficient vector finite element method for nonlinear electromagnetic modelling, 2006) . Later I can discuss with you to improve the efficiency  and other approches to linearise the problem.
Thanks in advance.
Best wishes
asad anees

**Comment by tzanio** [comment]
Hi @Asadanees,

If the mass matrix is on the Nedelec element space (the electric field), then you just need `VectorFEMassIntegrator`, see 

https://github.com/mfem/mfem/blob/1569425857975d7c961b97d09c300da09aad9f82/miniapps/electromagnetics/volta_solver.cpp#L113-L114

Note that this is not a 3x3 block matrix, but rather a matrix on the (scalar, edge-associated) Nedelec dofs.

Tzanio

**Comment by Asadanees** [comment]
Dear @tzanio and @mlstowell  @acfisher 
Might be , My question is not clear.  I will 
Dear @tzanio, I have implemented in the same way in linear and nonlinear Maxwell equations(nonliear Optics).
I have developed  mass matrix when the nonlinearity have the type 
( ε + 3 chi | E|² ) dE/dt , 
Whree E is a Electric Field and chi is scalar.
 My code is nice above type of nonlinearity.
Now I want to extend this code general type of non-linearity
I did not know how to contract the mass matrix at each time step when A is 3x3 Matrix
{(ε + 3 chi | E|²  I + 2 Chi ^3 A)} dE/dt
how we get this term above term

If D= (ε + Chi^3 |E|^2)E
when we take the derivative with respect to E 
dD/dT = {(ε + 3 chi | E|²  I + 2 Chi ^3 A)} dE/dt

I is 3 x 3 unit Matrix (Identity unit 3x3 matrix)
how we get |E|^2 and A
|E|^2 = E . E^T= E1^2+E2^2+E3^2= InnerProduct(E,E)
E= (E1, E2 , E3)
A= E^T E, then we get the 3x3 matrix A
A is 3x3 matrix , the entries of matrix are given below
Chi is a scalar. 

        E1.E1     E1.E2     E1.E3         ... first row of A
 A =  E1.E2     E2.E2     E2.E3         ...second row of A
        E1.E3      E3.E2     E3.E3        ...third row of A

IF we use this technique  then there is no need to update any matrix on the right hand side at each time steps.
We have to only update only Left hand side at each time step. we have to construct the mass matirix with weight is also Matrix
As I understand 
In the mfem-3.3 in fem-> bilintig.hpp ->class class VectorFEMassIntegrator 
has Constructor with weighted MatrixCoefficient.
private:
   MatrixCoefficient *MQ;
public:
 VectorFEMassIntegrator(MatrixCoefficient &mq) { Init(NULL, NULL, &mq); } 
I did not how to construct implemented the above code 
how to a mass Matrix in the above case.
Why am I interested in this above way? Because I have find the analytical solution and error estimate in this way
As I already said that I will use the approach @acfisher (an efficient vector finite element method for nonlinear electromagnetic modelling, 2006) . Here we need update both left a right hand side at each time step. 
**Later, I want to @extend  this code for Unstructured Mesh (both linear and nonlinear Maxwell equations) and add @parallel Time intergeneration. and it is property of LLNL(mfem)** 
I am thankful to @mlstowell for you help and contribute your name in my publications till December "Time domain finite element method  for nonlinear Optics. Before submitted to IEEE, I will send you full length paper to review. 
Could you please help me to speed up our work for Unstructured and  parallel time integration.
Feel free to if you have any question @tzanio , @mlstowell and @acfisher.
Thanks in advance.
Best wishes
asad anees

**Comment by Asadanees** [comment]
Dear @tzanio, @mlstowell and @acfisher .
Is it possible ? Could we  develop a mass matrix at each time step in the below  case?
** {(ε + 3 chi | E |^² *I* + 2 Chi ^3 A)} dE/dt**   for this case

# Chi and  **ε ** are scalars.
*I* is 3 x 3 **unit Matrix**  (Identity unit 3x3 matrix)
how we get |E|^2 and  **A**
|E|^2 = E . E^T= E1^2+E2^2+E3^2= InnerProduct(E,E)
**E= (E1, E2 , E3)^T** 
**A= E^T E**,    then we get the 3x3 matrix A
# **A**  is 3x3 matrix , the entries of matrix are given below


           E1.E1      E1.E2       E1.E3         **.. first row of A**
    A =  E1.E2      E2.E2       E2.E3         **... second row of A**
           E1.E3       E3.E2       E3.E3         **...third row of A**

Thanks,
Asad Anees

**Comment by acfisher** [comment]
Hi Asad,

If you plan to use time lagged values for E you should be able to do this by computing  {(εI + 3 chi | E |^² I + 2 Chi ^3 A)} at every time step and passing it into the VectorFEMassIntegrator as a matrix coefficient. There are lots of different ways you could compute this matrix.  Probably the easiest would be to project the values of E to the cell centers and use those as averages for the entire element giving you a constant matrix across the element to pass into the bilinear form.  You could definitely get fancier to get higher order solutions.

**Comment by Asadanees** [comment]
Thanks @acfisher   @mlstowell, @tzanio  .
Now I am closing this issue. Later I will reopen for pseudo code.
Best wishes .
